include/boost/url/detail/impl/remove_dot_segments.ipp

```diff
@@ -0,0 +1,518 @@
+    // - E leaves "/" or happens at the end
+    while (!s.empty())
+    {
+        if (s.starts_with("../"))
+        {
+            if (!remove_unmatched)
+            {
+                append(s.substr(0, 3));
+            }
+            s.remove_prefix(3);
+            continue;
+        }
+        if (!s.starts_with("./"))
+            break;
+        s.remove_prefix(2);
+    }
+
+    // D.  if the input buffer consists only
+    // of "." or "..", then remove that from
+    // the input buffer; otherwise,
+    // Rule D can only happen after A is consumed:
+    // - B and C write "/" to the output
+    // - D can only happen at the end
+    // - E leaves "/" or happens at the end
+    if (s == "." || s == "..")
+    {
+        if (!remove_unmatched && s == "..")
+        {
+            append(s);
+        }
+        s = {};
+    }
+
+    // 2. While the input buffer is not empty,
+    // loop as follows:
+    while (!s.empty())
+    {
+        // B.  if the input buffer begins with a
+        // prefix of "/./" or "/.", where "." is
+        // a complete path segment, then replace
+        // that prefix with "/" in the input
+        // buffer; otherwise,
+        if (s.starts_with("/./"))
+        {
+            s.remove_prefix(2);
+            continue;
+        }
+        if (s == "/.")
+        {
+            // equivalent to replacing s with '/'
+            // and executing the next iteration
+            append(s.substr(0, 1));
+            s.remove_prefix(2);
+            continue;
+        }
+
+        // C. if the input buffer begins with a
+        // prefix of "/../" or "/..", where ".."
+        // is a complete path segment, then
+        // replace that prefix with "/" in the
+        // input buffer and remove the last
+        // segment and its preceding "/"
+        // (if any) from the output buffer;
+        // otherwise,
+        if (s.starts_with("/../"))
+        {
+            std::size_t p = find_last_slash();
+            if (p != string_view::npos)
+            {
+                // "erase" [p, end]
+                dest = dest0 + p;
+            }
+            else if (!remove_unmatched)
+            {
+                append(s.substr(0, 3));
+            }
+            s.remove_prefix(3);
+            continue;
+        }
+        if (s == "/..")
+        {
+            std::size_t p = find_last_slash();
+            if (p != string_view::npos)
+            {
+                // erase [p, end]
+                dest = dest0 + p;
+                // equivalent to replacing s with '/'
+                // and executing the next iteration.
+                // this is the only point that would
+                // require input memory allocations
+                // in remove_dot_segments
+                append(s.substr(0, 1));
+            }
+            else if (remove_unmatched)
+            {
+                append(s.substr(0, 1));
+            }
+            else if (!remove_unmatched)
```

Comment:
You are testing `remove_unmatched` and then testing `! remove_unmatched`... and these lines are not covered.